### ISSUE TITLE
Case sensitivity bug with user.load and OAuth 

### ISSUE THREAD

**Issue by byjrack** [astuto/astuto#200]
_Using the public Docker Image and compose with an additional TLS termination container via nginx_

So the first login works great using my OAuth config, but the 2nd was just dropping back to "Login" with no errors.

Looking at production.log I can see a flow looking like
```
D, [2023-02-16T12:58:01.330500 #1059] DEBUG -- : [4967213a-aa7f-4f52-a266-383136150051]   User Load (0.7ms)  SELECT "users".* FROM "users" WHERE "users"."tenant_id" = $1 AND "users"."email" = $2 LIMIT $3  [["tenant_id", 1], ["email", "John.Doe@EXAMPLE.COM"], ["LIMIT", 1]]
D, [2023-02-16T12:58:01.452965 #1059] DEBUG -- : [4967213a-aa7f-4f52-a266-383136150051]    (0.3ms)  BEGIN
D, [2023-02-16T12:58:01.453711 #1059] DEBUG -- : [4967213a-aa7f-4f52-a266-383136150051]   User Exists? (0.5ms)  SELECT 1 AS one FROM "users" WHERE LOWER("users"."email") = LOWER($1) AND "users"."tenant_id" = $2 LIMIT $3  [["email", "john.doe@example.com"], ["tenant_id", 1], ["LIMIT", 1]]
D, [2023-02-16T12:58:01.455173 #1059] DEBUG -- : [4967213a-aa7f-4f52-a266-383136150051]    (0.3ms)  ROLLBACK
```

I checked in the users table and the value was stored in lowercase.  So the user.load code actually returns null even though the exists returns 1.  I went into the table and updated the email field to match the content of the OAuth email field and login was functional again.  If the user.create only stores lowercase than the user.load needs to cast to lower any input both from the UI (which it may already) and from an OAuth result.  I need to spend some time in the code on the persistence model to see if I can pick out what needs to be patched, but figured I would get the Issue opened as folks may know exactly where the code exists.

**Comment by byjrack** [comment]
And if you change the User (adjust Role) than the lower case is put back into the Db and needs to be created.

**Comment by riggraz** [comment]
Hi byjrack, thanks for reaching out!

Yes, I need to look into this, but your analysis is likely correct. It should be an easy fix.

I'll let you know when it's done. Thanks again!

**Comment by byjrack** [comment]
I have a few gates internally for PRs, but I think I tracked down the fix.

https://github.com/astuto/astuto/blob/main/app/workflows/OAuthSignInUser.rb#L35

I patched to a simple email.downcase in the find_by and restarted rails and login/logout seems to be working.  Never spent anytime in rails/activerecord so there may be a far more elegant fix.

**Comment by riggraz** [comment]
This is strange, because a setting in the user model should always lowercase email before performing queries. I'll investigate about this behavior. In the meanwhile, can you confirm that OAuth now works correctly your fix? Thank you!

**Comment by byjrack** [comment]
Far from a Ruby or ActiveRecord expert, but it looks like find_by isn't overloaded in the model and by default is case sensitive.  I did see that the model validates insensitive which had me going down the wrong path for a bit, but kept hunting.  Some folks use different built ins like where and LIKE statements, but since you persist the email as lower case in the model.create it was easy to just set the input to lower so the generated SQL equal test would function as expected.  The email param comes from the OAuth response and is not forced to lower before the search, but could be from line 30 onwards.  I actually was a bit surprised there wasn't a param for find_by that would control search sensitivity, but its a generic function for all the types so I get it.

And yeah I have not had any issues with session management since manually patching.  Still just a PoC so have to make sure to revisit if the container gets restarted as I am not using a locally built one for the moment.

**Comment by riggraz** [comment]
I thought that every query on the User model would be case insensitive on the email because I set up the email as case insensitive in the [devise config](https://github.com/astuto/astuto/blob/main/config/initializers/devise.rb#L55) (Devise is the package used for authentican). However, you were right in saying that find_by isn't affected by that and is case sensitive.

I found a probably more idiomatic solution by using the Devise method `find_for_authentication`, which is indeed case insensitive. You can check out the fix on PR #202.

Please let me know if this fix works for you too, and I'll be happy to merge the PR!

**Comment by byjrack** [comment]
On the surface it sounds like a great fix.  I will do a quick patch early this week to confirm the result is the same.
